In [18]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from scipy import stats

### Module 2, Phase 1: The Theoretical Lecture (Evaluating Regression)

#### 1. The Building Block: The Residual

The most fundamental unit of error is the **residual**, often denoted as $e_i$. The residual is the error for a *single observation*.

$$
e_i = y_i - \hat{y}_i
$$

* $y_i$: The $i$-th **true value** (the "correct answer" from our dataset).
* $\hat{y}_i$: The $i$-th **predicted value** (the "model's guess").

If our model predicts perfectly, $e_i = 0$. If it under-predicts, the residual is positive. If it over-predicts, the residual is negative.

Our goal is to find a model $\hat{f}$ that makes these residuals as small as possible, on average, across all our data points. The different ways we "average" these residuals give us our different metrics.

#### 2. The Toy Dataset: A Manual Calculation

To understand these metrics, we will *not* use a large dataset. We will use a tiny, 5-point dataset and calculate everything by hand, step-by-step.

Let's say we have 5 data points:
* **True Values ($Y$):** $\{2, 4, 5, 4, 5\}$
* **Our Model's Predictions ($\hat{Y}$):** $\{2.5, 3.5, 4.5, 4.5, 5.5\}$

First, let's calculate the residuals ($e_i$):
* $e_1 = 2 - 2.5 = -0.5$
* $e_2 = 4 - 3.5 = 0.5$
* $e_3 = 5 - 4.5 = 0.5$
* $e_4 = 4 - 4.5 = -0.5$
* $e_5 = 5 - 5.5 = -0.5$

Our residuals are: $\{-0.5, 0.5, 0.5, -0.5, -0.5\}$.

Notice that if we just sum them ($\sum e_i$), they add up to -0.5. They almost cancel out. This is why we can't just take a simple average of the residuals; we need to make them all positive first.

#### 3. Metric 1: Mean Absolute Error (MAE)

The Mean Absolute Error is the simplest and most direct way to average the errors. It answers the question: "On average, how far off is our prediction?"

**The Math:**
$$
MAE = \frac{1}{n} \sum_{i=1}^{n} |y_i - \hat{y}_i| = \frac{1}{n} \sum_{i=1}^{n} |e_i|
$$

* $|e_i|$: We take the **absolute value** of each residual. This turns all our negative errors into positive numbers.
* $\sum$: We sum them all up.
* $\frac{1}{n}$: We divide by $n$ (the number of data points) to get the mean.

**Manual Calculation:**
1.  **Residuals ($e_i$):** $\{-0.5, 0.5, 0.5, -0.5, -0.5\}$
2.  **Absolute Residuals ($|e_i|$):** $\{0.5, 0.5, 0.5, 0.5, 0.5\}$
3.  **Sum ($\sum |e_i|$):** $0.5 + 0.5 + 0.5 + 0.5 + 0.5 = 2.5$
4.  **Mean ($\frac{1}{n} \sum ...$):** $\frac{2.5}{5} = 0.5$

* **Result:** The **MAE is 0.5**.
* **Interpretation:** "On average, our model's prediction is 0.5 units away from the true value."
* **Key Property:** MAE is **robust to outliers**. If our last prediction was $50.5$ (an error of -45.5) instead of $5.5$, that one point would *linearly* increase the sum. An error of 10 adds 10 to the sum; an error of 100 adds 100. This makes MAE a good metric if your data has known, large, spurious outliers that you don't want to dominate the error metric.

#### 4. Metric 2: Mean Squared Error (MSE)

The Mean Squared Error is the most common metric used for *optimizing* regression models. It answers a slightly different question: "What is the average of the *squared* errors?"

**The Math:**
$$
MSE = \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2 = \frac{1}{n} \sum_{i=1}^{n} (e_i)^2
$$

* $(e_i)^2$: We **square** each residual. This *also* turns all negative errors positive.
* $\sum$: We sum them all up.
* $\frac{1}{n}$: We take the mean.

**Manual Calculation:**
1.  **Residuals ($e_i$):** $\{-0.5, 0.5, 0.5, -0.5, -0.5\}$
2.  **Squared Residuals ($e_i^2$):** $\{(-0.5)^2, (0.5)^2, (0.5)^2, (-0.5)^2, (-0.5)^2\} = \{0.25, 0.25, 0.25, 0.25, 0.25\}$
3.  **Sum ($\sum e_i^2$):** $0.25 + 0.25 + 0.25 + 0.25 + 0.25 = 1.25$
4.  **Mean ($\frac{1}{n} \sum ...$):** $\frac{1.25}{5} = 0.25$

* **Result:** The **MSE is 0.25**.
* **Interpretation:** "The average squared error is 0.25 units-squared." This is mathematically correct but not very interpretable, as the units are now squared (e.g., "squared dollars").
* **Key Properties:**
    1.  **Punishes Large Errors:** This is the most important property. Squaring the error makes the penalty *exponential*. An error of 2 becomes 4. An error of 10 becomes 100. An error of 100 becomes 10,000. This makes MSE *highly sensitive to outliers*. A model optimized to minimize MSE will work very hard to avoid large errors, even at the expense of making more small errors.
    2.  **Mathematical Convenience:** The function $f(x) = x^2$ is smooth and differentiable everywhere. The absolute value function $f(x) = |x|$ has a sharp "V" point at 0, which is not differentiable. This mathematical property makes MSE the default cost function for optimization algorithms that use calculus, like **Gradient Descent**.
    3.  **Theoretical Link:** Minimizing MSE is equivalent to finding the Maximum Likelihood Estimate (MLE) for a linear regression model *if* we assume the errors ($\epsilon$) are normally distributed.

#### 5. Metric 3: Root Mean Squared Error (RMSE)

The RMSE is the "fix" for MSE's interpretability problem. It is the most common metric for *reporting* regression error.

**The Math:**
$$
RMSE = \sqrt{MSE} = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}
$$

* It is simply the **square root of the MSE**.

**Manual Calculation:**
1.  **MSE:** We already found this is 0.25.
2.  **Square Root ($\sqrt{MSE}$):** $\sqrt{0.25} = 0.5$

* **Result:** The **RMSE is 0.5**.
* **Interpretation:** "On average, the magnitude of our model's error is 0.5 units." This is now in the same units as $Y$ (e.g., "dollars," not "squared dollars").
* **Properties:** RMSE gives you the best of both worlds:
    1.  It retains the *mathematical properties* of MSE (punishes large errors, used in optimization).
    2.  It has the *interpretable units* of MAE.
* **RMSE vs. MAE:** The RMSE will *always* be greater than or equal to the MAE. The only time they are equal is when all errors are identical. The larger the gap between RMSE and MAE, the more it suggests your model has some *large* errors that the RMSE is picking up on. In our toy example, all errors were identical (0.5), so $RMSE = MAE = 0.5$.

#### 6. Metric 4: R-Squared ($R^2$, Coefficient of Determination)

MAE, MSE, and RMSE are **absolute metrics**. An RMSE of 10,000 is terrible for predicting exam scores (0-100) but fantastic for predicting housing prices ($100k - $10M).

We need a **relative metric** that tells us how "good" our model is on a standardized scale, like a percentage. $R^2$ does this by answering the question: **"How much better is our model than a simple, naive baseline model?"**

The *most* naive (but sensible) baseline model is one that just predicts the *average* of $Y$ for every single prediction, ignoring $X$ entirely.

**The Math:**
$$
R^2 = 1 - \frac{\sum (y_i - \hat{y}_i)^2}{\sum (y_i - \bar{y})^2} = 1 - \frac{\text{Sum of Squared Residuals (SSR)}}{\text{Total Sum of Squares (SST)}}
$$

Let's break this down:
* $\sum (y_i - \hat{y}_i)^2$: This is the **Sum of Squared Residuals (SSR)**. It's the total squared error of *our* model. We already calculated this in the MSE section: it was 1.25.
* $\bar{y}$: This is the **mean of the true values $Y$**. This is the prediction of our *naive baseline model*.
    * For our data $Y = \{2, 4, 5, 4, 5\}$, $\bar{y} = \frac{2+4+5+4+5}{5} = \frac{20}{5} = 4$.
* $\sum (y_i - \bar{y})^2$: This is the **Total Sum of Squares (SST)**. It's the total squared error that the *naive mean model* would make. It represents the *total variance* in the $Y$ variable.

**Manual Calculation (for $R^2$):**
1.  **Calculate $\bar{y}$:** $\bar{y} = 4$.
2.  **Calculate SSR:** $\sum (y_i - \hat{y}_i)^2 = 1.25$ (from our MSE calculation).
3.  **Calculate SST:** $\sum (y_i - \bar{y})^2$
    * $e_1 = (2 - 4)^2 = (-2)^2 = 4$
    * $e_2 = (4 - 4)^2 = (0)^2 = 0$
    * $e_3 = (5 - 4)^2 = (1)^2 = 1$
    * $e_4 = (4 - 4)^2 = (0)^2 = 0$
    * $e_5 = (5 - 4)^2 = (1)^2 = 1$
    * **SST = $4 + 0 + 1 + 0 + 1 = 6$**
4.  **Calculate $R^2$:**
    * $R^2 = 1 - \frac{SSR}{SST} = 1 - \frac{1.25}{6.0}$
    * $R^2 = 1 - 0.2083...$
    * $R^2 \approx 0.7917$

* **Result:** The **$R^2$ is approximately 0.792**.
* **Interpretation:** "Our model **explains 79.2% of the variance** in the $Y$ variable."
* **Properties:**
    * $R^2 = 1.0$: Perfect model. $SSR=0$.
    * $R^2 = 0.0$: Our model is no better than just predicting the mean. $SSR = SST$.
    * $R^2 < 0.0$: Our model is *worse* than just predicting the mean. $SSR > SST$. This can easily happen on a **test set** if your model is badly overfit or just plain wrong.

**A Critical Warning:** $R^2$ on the *training set* has a major flaw: it will *always* increase (or stay the same) every time you add a new feature to the model, even if that feature is useless. This can be misleading. This is why we either:
1.  Only judge $R^2$ on the **test set**.
2.  Use **Adjusted $R^2$**, which is a modified version that penalizes the score for adding features that don't actually improve the model.

## Regression Evaluation Metrics Summary

| Metric | Formula | Interpretation | Key Properties |
|--------|----------|-----------------|----------------|
| **MAE**<br>(Mean Absolute Error) | $\frac{1}{n}\sum \|e_i\|$ | Average absolute prediction error | • **Robust to outliers**<br>• Same units as $Y$<br>• Easy to interpret |
| **MSE**<br>(Mean Squared Error) | $\frac{1}{n}\sum (e_i)^2$ | Average squared prediction error | **Sensitive to outliers**<br>• $Y^2$ units<br>• **Differentiable**<br>• Default for optimization |
| **RMSE**<br>(Root Mean Squared Error) | $\sqrt{\frac{1}{n}\sum (e_i)^2}$ | Standard deviation of residuals | • **Most popular**<br>• Same units as $Y$<br>• Sensitive to large errors |
| **$R^2$**<br>(R-squared) | $1 - \frac{SSR}{SST}$ | Proportion of variance explained | • **Relative scale (0-1)**<br>• Percentage interpretation<br>• **Variance explained** |

### Key Relationships:
- **RMSE** = $\sqrt{\text{MSE}}$
- **$R^2$** = $1 - \frac{\text{SSE}}{\text{SST}}$ where SSE = $\sum(e_i)^2$
- **MSE** penalizes large errors more severely than MAE due to squaring

### When to Use Which:
- **MAE**: When outliers should not be heavily penalized
- **MSE/RMSE**: Default choice for most regression problems
- **RMSE**: Preferred when interpretability in original units matters
- **$R^2$**: For understanding explanatory power relative to simple mean model

### Module 2, Phase 2: The Applied Code Lecture (Evaluating Regression)

Our workflow picks up *after* we have already trained a model. The goal here is to evaluate the predictions that model has made on the **test set**.

**1. Setup: Re-creating the Model and Predictions**

First, we'll quickly reload the data, split it, and train a basic `LinearRegression` model. This gives us the `y_test` (true values) and `y_pred` (model's predictions) that we need.

In [19]:
from sklearn.datasets import fetch_california_housing
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

# 1. Load Data
housing = fetch_california_housing()
X, y = housing.data, housing.target

# 2. Split Data
# We use the same split as before for consistency
X_train, X_test, y_train, y_test = train_test_split(
	X, y, test_size=0.2, random_state=42
)

# 3. Fit a Model
# We'll use a simple Linear Regression as our example model
model = LinearRegression()
model.fit(X_train, y_train)

# 4. Get Predictions
# This is the crucial step: we get predictions on the *unseen* test set
y_pred = model.predict(X_test)

# We now have the two arrays we need:
# y_test: The "true values" (our y_i)
# y_pred: The "model's guesses" (our ŷ_i)
print(f"First 5 True Values (y_test):   {np.round(y_test[:5], 2)}")
print(f"First 5 Predicted Values (y_pred): {np.round(y_pred[:5], 2)}")

First 5 True Values (y_test):   [0.48 0.46 5.   2.19 2.78]
First 5 Predicted Values (y_pred): [0.72 1.76 2.71 2.84 2.6 ]



**2. Implementing the Metrics with `sklearn.metrics`**

Now, we will calculate the four key metrics. The `sklearn.metrics` module provides functions that directly implement the formulas we discussed. The standard pattern for all of them is `function_name(y_true, y_pred)`.

In [20]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Metric 1: Mean Absolute Error (MAE)
# Formula: (1/n) * Σ|y_i - ŷ_i|
mae = mean_absolute_error(y_test, y_pred)
print("--- Mean Absolute Error (MAE) ---")
print(f"MAE: {mae:.4f}")

# Metric 2: Mean Squared Error (MSE)
# Formula: (1/n) * Σ(y_i - ŷ_i)²
mse = mean_squared_error(y_test, y_pred)
print("--- Mean Squared Error (MSE) ---")
print(f"MSE: {mse:.4f}")

# Metric 3: Root Mean Squared Error (RMSE)
# Formula: sqrt(MSE)
# Scikit-learn doesn't have a separate 'rmse' function
# because it's just the square root of MSE.
rmse = np.sqrt(mse)
print("--- Root Mean Squared Error (RMSE) ---")
print(f"RMSE: {rmse:.4f}")

# Metric 4: R-Squared (R²)
# Formula: 1 - (SSR / SST)
# SSR = Σ(y_i - ŷ_i)²
# SST = Σ(y_i - ȳ)²
r2 = r2_score(y_test, y_pred)
# You can also get this from the model's .score() method
r2_from_model = model.score(X_test, y_test)
print("--- R-Squared (R²) ---")
print(f"R² (from r2_score): {r2:.4f}")
print(f"R² (from model.score): {r2_from_model:.4f}")

--- Mean Absolute Error (MAE) ---
MAE: 0.5332
--- Mean Squared Error (MSE) ---
MSE: 0.5559
--- Root Mean Squared Error (RMSE) ---
RMSE: 0.7456
--- R-Squared (R²) ---
R² (from r2_score): 0.5758
R² (from model.score): 0.5758


**3. Interpretation: What Do These Numbers Mean?**

Let's assume our code output the following:

  * **MAE:** 0.5344
  * **MSE:** 0.5559
  * **RMSE:** 0.7456
  * **R²:** 0.5758

Here is how to interpret this in the context of the California Housing data, where the target `y` is median house value in units of $100,000s.

  * **MAE: 0.5344**

      * **Interpretation:** "On average, our model's prediction of a district's median house value is **$53,440** off from the true value." (0.5344 \* $100,000). This is the most direct and interpretable error value.

  * **RMSE: 0.7456**

      * **Interpretation:** This is another measure of average error magnitude, also in units of $100,000s. We would say the root mean squared error is **$74,560\*\*.
      * **RMSE vs. MAE:** Notice that our RMSE (0.7456) is significantly larger than our MAE (0.5344). As we discussed in Phase 1, this implies that our model is making *some* very large errors. The squaring process in RMSE penalizes these large errors heavily, which inflates the RMSE value more than the MAE value. This tells us our model's performance is not uniform; it has some significant misses.

  * **MSE: 0.5559**

      * **Interpretation:** "The average squared error is 0.5559." This number is mathematically vital for optimization but is not human-interpretable (its units are "squared $100,000s"). We report RMSE instead.

  * **R²: 0.5758**

      * **Interpretation:** "Our model **explains 57.6% of the variance** in median house prices on the test set." This means that just under 60% of the reason *why* some districts are more expensive than others is captured by the features in our simple linear model. The other 42.4% is due to factors not in our model or the model's inability to capture the relationships (i.e., the *residuals*). This is a mediocre, but not terrible, score, suggesting there is significant room for improvement with a more complex model.


### Module 2, Phase 2.1: Implementing Metrics from Scratch

We will use the same `y_test` (the true values) and `y_pred` (the model's predictions) that we generated in the setup for Phase 2, from the California Housing data. Our goal is to create our own functions that replicate `sklearn.metrics`.

Our inputs will be two NumPy arrays: `y_true` and `y_pred`.

**1. Mean Absolute Error (MAE) from Scratch**

  * **The Math:** $MAE = \frac{1}{n} \sum_{i=1}^{n} |y_i - \hat{y}_i|$
  * **The Logic:**
    1.  Calculate all residuals: `y_true - y_pred`.
    2.  Take the absolute value of each residual: `np.abs(...)`.
    3.  Take the mean of all absolute residuals: `np.mean(...)`. `np.mean` handles both the sum ($\sum$) and the division ($\frac{1}{n}$) in one step.



In [21]:
def mae_from_scratch(y_true: np.ndarray, y_pred: np.ndarray) -> float:
	"""
	Calculate Mean Absolute Error (MAE) from scratch.

	MAE is the average of the absolute differences between predicted and true values.

	Parameters
	----------
	y_true : array-like of shape (n_samples,)
		Ground truth (correct) target values.
	y_pred : array-like of shape (n_samples,)
		Estimated target values.

	Returns
	-------
	mae : float
		Mean Absolute Error value.

	Notes
	-----
	Formula: MAE = (1/n) * Σ|y_true - y_pred|

	Examples
	--------
	>>> y_true = [3, -0.5, 2, 7]
	>>> y_pred = [2.5, 0.0, 2, 8]
	>>> mae_from_scratch(y_true, y_pred)
	0.5
	"""
	# Ensure inputs are NumPy arrays for vector operations
	y_true = np.asarray(y_true)
	y_pred = np.asarray(y_pred)

	# 1. Calculate all residuals (e_i = y_i - ŷ_i)
	residuals = y_true - y_pred

	# 2. Take the absolute value of each residual (|e_i|)
	absolute_residuals = np.abs(residuals)

	# 3. Take the mean of all absolute residuals ( (1/n) * Σ|e_i| )
	mae = np.mean(absolute_residuals)

	return mae

**2. Mean Squared Error (MSE) from Scratch**

  * **The Math:** $MSE = \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2$
  * **The Logic:**
    1.  Calculate all residuals: `y_true - y_pred`.
    2.  Square each residual: `... ** 2` or `np.square(...)`.
    3.  Take the mean of all squared residuals: `np.mean(...)`.

In [22]:
def mse_from_scratch(y_true: np.ndarray, y_pred: np.ndarray) -> float:
	"""
	Calculate Mean Squared Error (MSE) from scratch.

	MSE is the average of the squared differences between predicted and true values.

	Parameters
	----------
	y_true : array-like of shape (n_samples,)
		Ground truth (correct) target values.
	y_pred : array-like of shape (n_samples,)
		Estimated target values.

	Returns
	-------
	mse : float
		Mean Squared Error value.

	Notes
	-----
	Formula: MSE = (1/n) * Σ(y_true - y_pred)²

	MSE is sensitive to outliers due to the squaring operation.
	Larger errors contribute disproportionately more to the total error.

	Examples
	--------
	>>> y_true = [3, -0.5, 2, 7]
	>>> y_pred = [2.5, 0.0, 2, 8]
	>>> mse_from_scratch(y_true, y_pred)
	0.375
	"""
	# Ensure inputs are NumPy arrays for vector operations
	y_true = np.asarray(y_true)
	y_pred = np.asarray(y_pred)

	# 1. Calculate all residuals (e_i = y_i - ŷ_i)
	residuals = y_true - y_pred

	# 2. Square each residual (e_i²)
	squared_residuals = residuals**2
	# Alternatively: np.square(residuals)

	# 3. Take the mean of all squared residuals ( (1/n) * Σ(e_i)² )
	mse = np.mean(squared_residuals)

	return mse

**3. Root Mean Squared Error (RMSE) from Scratch**

  * **The Math:** $RMSE = \sqrt{MSE}$
  * **The Logic:**
    1.  Calculate the MSE using our function from step 2.
    2.  Take the square root of the result: `np.sqrt(...)`.

In [23]:
def rmse_from_scratch(y_true: np.ndarray, y_pred: np.ndarray) -> float:
	"""
	Calculate Root Mean Squared Error (RMSE) from scratch.

	RMSE is the square root of the average of the squared differences
	between predicted and true values. It provides error measurement
	in the same units as the target variable.

	Parameters
	----------
	y_true : array-like of shape (n_samples,)
		Ground truth (correct) target values.
	y_pred : array-like of shape (n_samples,)
		Estimated target values.

	Returns
	-------
	rmse : float
		Root Mean Squared Error value.

	Notes
	-----
	Formula: RMSE = √[(1/n) * Σ(y_true - y_pred)²]

	RMSE is sensitive to outliers due to the squaring operation.
	It's one of the most commonly used regression evaluation metrics.

	Examples
	--------
	>>> y_true = [3, -0.5, 2, 7]
	>>> y_pred = [2.5, 0.0, 2, 8]
	>>> rmse_from_scratch(y_true, y_pred)
	0.6123724356957945
	"""
	# 1. Calculate MSE using our existing function
	mse = mse_from_scratch(y_true, y_pred)

	# 2. Take the square root to get RMSE
	rmse = np.sqrt(mse)

	return rmse

**4. R-Squared ($R^2$) from Scratch**

  * **The Math:** $R^2 = 1 - \frac{SSR}{SST} = 1 - \frac{\sum (y_i - \hat{y}_i)^2}{\sum (y_i - \bar{y})^2}$
  * **The Logic:** This is the most complex one. We need two parts.
    1.  **Calculate SSR (Sum of Squared Residuals):** This is the numerator. It's the sum of our squared errors.
    2.  **Calculate SST (Total Sum of Squares):** This is the denominator. It's the sum of the squared differences between the *true values* and the *mean of the true values*.
    3.  Apply the final formula.

In [24]:
def r2_from_scratch(y_true: np.ndarray, y_pred: np.ndarray) -> float:
	"""
	Calculate R-squared (R²) from scratch.

	R-squared measures the proportion of variance in the dependent variable
	that is predictable from the independent variables. It represents how
	much better the model is compared to simply predicting the mean.

	Parameters
	----------
	y_true : array-like of shape (n_samples,)
		Ground truth (correct) target values.
	y_pred : array-like of shape (n_samples,)
		Estimated target values.

	Returns
	-------
	r2 : float
		R-squared value between 0 and 1 (can be negative for poor models).

	Notes
	-----
	Formula: R² = 1 - (SSR / SST)
	- SSR (Sum of Squared Residuals) = Σ(y_true - y_pred)²
	- SST (Total Sum of Squares) = Σ(y_true - mean(y_true))²

	Interpretation:
	- R² = 1: Perfect prediction
	- R² = 0: Model performs no better than mean prediction
	- R² < 0: Model performs worse than mean prediction

	Examples
	--------
	>>> y_true = [3, -0.5, 2, 7]
	>>> y_pred = [2.5, 0.0, 2, 8]
	>>> r2_from_scratch(y_true, y_pred)
	0.9486081370449679
	"""
	# Ensure inputs are NumPy arrays for vector operations
	y_true = np.asarray(y_true)
	y_pred = np.asarray(y_pred)

	# 1. Calculate SSR (Sum of Squared Residuals)
	# This is the numerator of the fraction
	residuals = y_true - y_pred
	ssr = np.sum(residuals**2)

	# 2. Calculate SST (Total Sum of Squares)
	# This is the denominator of the fraction
	# First, find the mean of the true values (ȳ)
	mean_y_true = np.mean(y_true)
	# Then, find the squared differences from the mean
	total_variance_residuals = y_true - mean_y_true
	sst = np.sum(total_variance_residuals**2)

	# 3. Apply the R² formula
	# (Handle edge case where SST is zero to avoid division by zero)
	if sst == 0:
		# This means y_true is a constant value.
		# If predictions are also perfect (ssr=0), R² is 1.0
		# If predictions are not perfect (ssr>0), R² is undefined or 0 (model is no help)
		return 1.0 if ssr == 0 else 0.0

	r2 = 1 - (ssr / sst)

	return r2

#### Verification: Running Our Functions

Now, let's use the `y_test` and `y_pred` from our Phase 2 setup (the California Housing data) and compare our scratch functions directly against the `scikit-learn` versions.

In [25]:
# Setup from Phase 2 (for context)
# X_train, X_test, y_train, y_test = ...
# model = LinearRegression()
# model.fit(X_train, y_train)
# y_pred = model.predict(X_test)
# y_test and y_pred are now our NumPy arrays.

print("--- Verification of Scratch Implementations ---")

mae_scratch = mae_from_scratch(y_test, y_pred)
print(f"MAE (from scratch):  {mae_scratch:.4f}  (sklearn was 0.5344)")

mse_scratch = mse_from_scratch(y_test, y_pred)
print(f"MSE (from scratch):  {mse_scratch:.4f}  (sklearn was 0.5559)")

rmse_scratch = rmse_from_scratch(y_test, y_pred)
print(f"RMSE (from scratch): {rmse_scratch:.4f}  (sklearn was 0.7456)")

r2_scratch = r2_from_scratch(y_test, y_pred)
print(f"R² (from scratch):   {r2_scratch:.4f}  (sklearn was 0.5758)")

--- Verification of Scratch Implementations ---
MAE (from scratch):  0.5332  (sklearn was 0.5344)
MSE (from scratch):  0.5559  (sklearn was 0.5559)
RMSE (from scratch): 0.7456  (sklearn was 0.7456)
R² (from scratch):   0.5758  (sklearn was 0.5758)


In [26]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

print("Direct Comparison:")
print(f"MAE match: {np.isclose(mae_scratch, mean_absolute_error(y_test, y_pred))}")
print(f"MSE match: {np.isclose(mse_scratch, mean_squared_error(y_test, y_pred))}")
print(
	f"RMSE match: {np.isclose(rmse_scratch, np.sqrt(mean_squared_error(y_test, y_pred)))}"
)
print(f"R2 match: {np.isclose(r2_scratch, r2_score(y_test, y_pred))}")

Direct Comparison:
MAE match: True
MSE match: True
RMSE match: True
R2 match: True


### Module 2, Phase 3: Applied Practicum - Calculating & Interpreting Regression Metrics

#### Dataset: Auto MPG

For this practicum, we will use the "Auto MPG" dataset. It is a classic dataset from the UCI repository, used to predict the fuel efficiency (in miles per gallon, MPG) of cars from the 1970s and 80s based on attributes like horsepower, weight, and cylinders.

It is a clean, intuitive regression problem.

#### Starter Code (Minimal Loader & Model)

The purpose of this practicum is to *evaluate predictions*, so I am providing the minimal code to load the data, clean it (it has a few missing `horsepower` values), split it, and train a single baseline `LinearRegression` model.

This will give you the two essential arrays you need to perform the tasks: `y_test` (the true values) and `y_pred` (the model's predictions).

In [27]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

# 1. Load Data
mpg = sns.load_dataset("mpg")

# 2. Minimal Cleaning
mpg_clean = mpg.dropna()
print(f"Original data shape: {mpg.shape}")
print(f"Cleaned data shape: {mpg_clean.shape}")

# 3. Define X and y
features = ["horsepower", "weight", "cylinders"]
target = "mpg"

X = mpg_clean[features]
y = mpg_clean[target]

# 4. Split Data
X_train, X_test, y_train, y_test = train_test_split(
	X, y, test_size=0.2, random_state=42
)

# 5. Fit a Baseline Model
baseline_model = LinearRegression()
baseline_model.fit(X_train, y_train)

# 6. Get Predictions
y_pred = baseline_model.predict(X_test)
# y_test is the corresponding array of true values.

print("Model fitting complete. You now have 'y_test' and 'y_pred'.")
print(f"Example True Values (y_test):   {np.round(y_test.values[:5], 2)}")
print(f"Example Predicted Values (y_pred): {np.round(y_pred[:5], 2)}")

Original data shape: (398, 9)
Cleaned data shape: (392, 9)
Model fitting complete. You now have 'y_test' and 'y_pred'.
Example True Values (y_test):   [26.  21.6 36.1 26.  27. ]
Example Predicted Values (y_pred): [30.13 24.69 32.65 31.34 26.09]


#### Tasks

You now have `y_test` (a Pandas Series, which you can use as a NumPy array with `y_test.values`) and `y_pred` (a NumPy array). Perform the following tasks.

**Task 1: Calculate Metrics using `scikit-learn`**

1.  Import `mean_absolute_error`, `mean_squared_error`, and `r2_score` from `sklearn.metrics`.
2.  Calculate the MAE, MSE, RMSE, and $R^2$ for the model's predictions (`y_pred`) against the true values (`y_test`).
3.  **Deliverable:** A code cell that prints out the four calculated metric values, clearly labeled.

In [28]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Absolute Error (sklearn): {mae:.4f}")
print(f"Mean Squared Error (sklearn): {mse:.4f}")
print(f"R² (sklearn): {r2:.4f}")

Mean Absolute Error (sklearn): 3.4588
Mean Squared Error (sklearn): 17.6772
R² (sklearn): 0.6537


**Task 2: Calculate Metrics from Scratch**

1.  Write your own Python functions, using only `NumPy`, to calculate each of the four metrics.
      * `def my_mae(y_true, y_pred)`
      * `def my_mse(y_true, y_pred)`
      * `def my_rmse(y_true, y_pred)` (This can call your `my_mse` function)
      * `def my_r2_score(y_true, y_pred)`
2.  Follow the mathematical formulas we discussed in Phase 1 and 2.1. Pay close attention to the formula for $R^2$ (it requires calculating both $SSR$ and $SST$).
3.  **Deliverable:** A code cell containing the definitions of these four functions.

### Mean Absolute Error (MAE)

$$MAE = \frac{1}{n} \sum_{i=1}^{n} |y_i - \hat{y}_i|$$



In [29]:
def my_mae(y_true: np.ndarray, y_pred: np.ndarray) -> float:
	"""
	Calculate Mean Absolute Error (MAE) from scratch.

	MAE is the average of the absolute differences between predicted and true values.

	Parameters
	----------
	y_true : array-like of shape (n_samples,)
		Ground truth (correct) target values.
	y_pred : array-like of shape (n_samples,)
		Estimated target values.

	Returns
	-------
	float
		Mean Absolute Error value.

	Notes
	-----
	Formula: MAE = (1/n) * Σ|y_true - y_pred|

	Examples
	--------
	>>> y_true = [3, -0.5, 2, 7]
	>>> y_pred = [2.5, 0.0, 2, 8]
	>>> my_mae(y_true, y_pred)
	0.5

	>>> y_true = [10, 20, 30]
	>>> y_pred = [12, 18, 35]
	>>> my_mae(y_true, y_pred)
	3.0
	"""
	y_true = np.asarray(y_true)
	y_pred = np.asarray(y_pred)

	# Calculate residuals (differences between true and predicted values)
	residuals = y_true - y_pred

	# Take absolute value of each residual
	absolute_residuals = np.abs(residuals)

	# Calculate mean of absolute residuals
	mae = np.mean(absolute_residuals)

	return mae

### Mean Squared Error (MSE)

$$MSE = \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2$$

In [30]:
def my_mse(y_true: np.ndarray, y_pred: np.ndarray) -> float:
	"""
	Calculate Mean Squared Error (MSE) from scratch.

	MSE is the average of the squared differences between predicted and true values.
	It is sensitive to outliers due to the squaring operation, which penalizes larger
	errors disproportionately more than smaller errors.

	Parameters
	----------
	y_true : array-like of shape (n_samples,)
		Ground truth (correct) target values.
	y_pred : array-like of shape (n_samples,)
		Estimated target values.

	Returns
	-------
	mse : float
		Mean Squared Error value.

	Notes
	-----
	Formula: MSE = (1/n) * Σ(y_true - y_pred)²

	MSE is commonly used as a loss function for regression models because:
	- It is differentiable everywhere (unlike MAE)
	- It heavily penalizes large errors
	- Minimizing MSE is equivalent to maximum likelihood estimation under
	  Gaussian noise assumptions

	Examples
	--------
	>>> y_true = np.array([3, -0.5, 2, 7])
	>>> y_pred = np.array([2.5, 0.0, 2, 8])
	>>> my_mse(y_true, y_pred)
	0.375

	>>> y_true = np.array([10, 20, 30])
	>>> y_pred = np.array([12, 18, 35])
	>>> my_mse(y_true, y_pred)
	13.0
	"""
	y_true = np.asarray(y_true)
	y_pred = np.asarray(y_pred)

	# Calculate residuals (differences between true and predicted values)
	residuals = y_true - y_pred

	# Square each residual to penalize larger errors more heavily
	resid_squared = residuals**2

	# Calculate mean of squared residuals
	mse = np.mean(resid_squared)

	return mse

### Root Mean Squared Error (RMSE)

$$RMSE = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2}$$

In [31]:
def my_rmse(y_true: np.ndarray, y_pred: np.ndarray) -> float:
	"""
	Calculate Root Mean Squared Error (RMSE) from scratch.

	RMSE is the square root of the average of the squared differences
	between predicted and true values. It provides error measurement
	in the same units as the target variable.

	Parameters
	----------
	y_true : array-like of shape (n_samples,)
		Ground truth (correct) target values.
	y_pred : array-like of shape (n_samples,)
		Estimated target values.

	Returns
	-------
	rmse : float
		Root Mean Squared Error value.

	Notes
	-----
	Formula: RMSE = √[(1/n) * Σ(y_true - y_pred)²]

	RMSE is sensitive to outliers due to the squaring operation.
	It's one of the most commonly used regression evaluation metrics.

	Examples
	--------
	>>> y_true = np.array([3, -0.5, 2, 7])
	>>> y_pred = np.array([2.5, 0.0, 2, 8])
	>>> my_rmse(y_true, y_pred)
	0.6123724356957945

	>>> y_true = np.array([10, 20, 30])
	>>> y_pred = np.array([12, 18, 35])
	>>> my_rmse(y_true, y_pred)
	3.605551275463989

	>>> # Using the Auto MPG dataset from the notebook
	>>> rmse_value = my_rmse(y_test.values, y_pred)
	>>> print(f"RMSE for Auto MPG model: {rmse_value:.4f}")
	RMSE for Auto MPG model: 4.2044
	"""
	y_true = np.asarray(y_true)
	y_pred = np.asarray(y_pred)

	# Calculate residuals (differences between true and predicted values)
	residuals = y_true - y_pred

	# Square each residual to penalize larger errors more heavily
	squared_residuals = residuals**2

	# Calculate mean of squared residuals (MSE)
	mse = np.mean(squared_residuals)

	# Take square root to get RMSE (back to original units)
	rmse = np.sqrt(mse)

	return rmse

### R-Squared (R²)

$$R^2 = 1 - \frac{\sum (y_i - \hat{y}_i)^2}{\sum (y_i - \bar{y})^2}$$

**Where:**
- $y_i$ = true values
- $\hat{y}_i$ = predicted values  
- $\bar{y}$ = mean of true values

In [32]:
def my_r2_score(y_true: np.ndarray, y_pred: np.ndarray) -> float:
	"""
	Calculate R-squared (R²) from scratch.

	R-squared measures the proportion of variance in the dependent variable
	that is predictable from the independent variables. It represents how
	much better the model is compared to simply predicting the mean.

	Parameters
	----------
	y_true : array-like of shape (n_samples,)
		Ground truth (correct) target values.
	y_pred : array-like of shape (n_samples,)
		Estimated target values.

	Returns
	-------
	r2 : float
		R-squared value between 0 and 1 (can be negative for poor models).

	Notes
	-----
	Formula: R² = 1 - (SSR / SST)
	- SSR (Sum of Squared Residuals) = Σ(y_true - y_pred)²
	- SST (Total Sum of Squares) = Σ(y_true - mean(y_true))²

	Interpretation:
	- R² = 1: Perfect prediction
	- R² = 0: Model performs no better than mean prediction
	- R² < 0: Model performs worse than mean prediction

	Examples
	--------
	>>> y_true = np.array([3, -0.5, 2, 7])
	>>> y_pred = np.array([2.5, 0.0, 2, 8])
	>>> my_r2_score(y_true, y_pred)
	0.9486081370449679

	>>> y_true = np.array([10, 20, 30])
	>>> y_pred = np.array([12, 18, 35])
	>>> my_r2_score(y_true, y_pred)
	0.75

	>>> # Perfect prediction
	>>> y_true = np.array([1, 2, 3])
	>>> y_pred = np.array([1, 2, 3])
	>>> my_r2_score(y_true, y_pred)
	1.0

	>>> # Worse than mean prediction
	>>> y_true = np.array([1, 2, 3])
	>>> y_pred = np.array([10, 10, 10])
	>>> my_r2_score(y_true, y_pred)
	-13.0
	"""
	y_true = np.asarray(y_true)
	y_pred = np.asarray(y_pred)

	# Calculate SSR (Sum of Squared Residuals)
	# This is the sum of squared differences between true and predicted values
	ssr = np.sum((y_true - y_pred) ** 2)

	# Calculate SST (Total Sum of Squares)
	# This is the sum of squared differences between true values and their mean
	mean_y_true = np.mean(y_true)
	sst = np.sum((y_true - mean_y_true) ** 2)

	# Handle edge case where SST is zero (all y_true values are the same)
	if sst == 0:
		# If all true values are identical, perfect prediction gives R² = 1
		# Otherwise, model is no better than predicting the constant
		return 1.0 if ssr == 0 else 0.0

	# Apply the R² formula: 1 - (SSR / SST)
	r2 = 1 - (ssr / sst)

	return r2

**Task 3: Compare and Verify**

1.  Run your four "scratch" functions from Task 2 using `y_test` and `y_pred`.
2.  **Deliverable:** A code cell that prints your scratch-calculated metrics. In a markdown cell, confirm whether your results match the `scikit-learn` results from Task 1.

In [33]:
mae = my_mae(y_test, y_pred)
mse = my_mse(y_test, y_pred)
rmse = my_rmse(y_test, y_pred)
r2 = my_r2_score(y_test, y_pred)

print(f"Mean Absolute Error (scratch): {mae:.4f}")
print(f"Mean Squared Error (scratch): {mse:.4f}")
print(f"Root Mean Squared Error (scratch): {rmse:.4f}")
print(f"R² (scratch): {r2:.4f}")

Mean Absolute Error (scratch): 3.4588
Mean Squared Error (scratch): 17.6772
Root Mean Squared Error (scratch): 4.2044
R² (scratch): 0.6537


In [34]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

print("Direct Comparison:")
print(f"MAE match: {np.isclose(mae, mean_absolute_error(y_test, y_pred))}")
print(f"MSE match: {np.isclose(mse, mean_squared_error(y_test, y_pred))}")
print(f"RMSE match: {np.isclose(rmse, np.sqrt(mean_squared_error(y_test, y_pred)))}")
print(f"R² match: {np.isclose(r2, r2_score(y_test, y_pred))}")

Direct Comparison:
MAE match: True
MSE match: True
RMSE match: True
R² match: True


**Task 4: Interpretation Report**

1.  Based on the (now verified) metric values, write a short report in a markdown cell.
2.  **Deliverable:** Your report must answer the following questions:
      * **MAE:** In plain English, what does the MAE value *mean*? (i.e., "On average, our model's prediction for..."). What are the *units* of this error?
      * **RMSE:** What is the RMSE value and what are its units?
      * **RMSE vs. MAE:** Is the RMSE larger or smaller than the MAE? What does the *difference* between them (or lack thereof) imply about the nature of the model's errors (e.g., are there many large errors, or are errors uniform)?
      * **$R^2$:** What is the $R^2$ value? In plain English, what does this percentage *mean* in the context of predicting "Miles Per Gallon"? (i.e., "Our model explains...").


#### MAE (Mean Absolute Error)

On average, our model’s predictions are off by **X miles per gallon** from the actual values.

**Units:**
Same as the target - here, **miles per gallon (MPG)**.

#### RMSE (Root Mean Squared Error)

RMSE is another measure of average error, but it **penalizes big mistakes more** than MAE does. It tells us the typical size of a prediction error, also in **miles per gallon**.

**Units:**
Same as the target - **MPG**.

#### RMSE vs. MAE

If **RMSE > MAE**, it means there are **some large errors** - a few predictions are much worse than the rest.
If **RMSE ≈ MAE**, errors are **fairly uniform** - no big outliers throwing things off.


#### ( $R^2$ ) (Coefficient of Determination)

( $R^2$ ) tells us how much of the variation in real MPG values our model can explain.
For example,

* ( $R^2$ = 0.85 ) → “Our model explains **85% of the variation** in Miles Per Gallon.”
* The remaining 15% comes from randomness or factors the model doesn’t capture.

In short:

* **MAE:** average miss, in MPG.
* **RMSE:** like MAE, but big misses hurt more.
* **RMSE vs MAE:** reveals whether errors are steady or spiky.
* **( $R^2$ ):** how much of the MPG story your model actually understands.
